mass_mailing_partner/models/mail_mass_mailing_contact.py

```diff
@@ -0,0 +1,75 @@
+# -*- coding: utf-8 -*-
+# See README.rst file on addon root folder for license details
+
+from openerp import models, fields, api, _
+
+
+class MailMassMailingContact(models.Model):
+    _inherit = 'mail.mass_mailing.contact'
+
+    partner_id = fields.Many2one(comodel_name='res.partner', string="Partner",
+                                 domain=[('email', '!=', False)])
+
+    _sql_constraints = [
+        ('partner_list_uniq', 'unique(partner_id, list_id)',
+         _('Partner already exists in this mailing list.'))
+    ]
+
+    @api.one
+    @api.onchange('partner_id')
+    def _onchange_partner(self):
+        if self.partner_id:
+            self.name = self.partner_id.name
+            self.email = self.partner_id.email
+
+    @api.model
```

Comment:
add `@api.returns('self', lambda x: x.id)` to downgrade correctly for legacy code